File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/AbstractProgressMediaView.java
package com.pr0gramm.app.ui.views.viewer;

import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.ProgressBar;
import android.widget.SeekBar;

import com.google.common.base.Optional;
import com.pr0gramm.app.R;
import com.pr0gramm.app.Settings;
import com.pr0gramm.app.util.AndroidUtility;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;

import static com.pr0gramm.app.util.AndroidUtility.endAction;

/**
 */
public abstract class AbstractProgressMediaView extends MediaView {
    private static final Logger logger = LoggerFactory.getLogger("AbstractProgressMediaView");

    @Inject
    Settings settings;

    boolean progressTouched = false;
    long lastUserInteraction = -1;

    private boolean progressEnabled = true;
    private boolean firstTimeProgressValue = true;

    private final SeekBar seekBarView;
    private final ProgressBar progressView;

    AbstractProgressMediaView(Config config, @LayoutRes Integer layoutId) {
        super(config, layoutId);

        progressView = (ProgressBar) LayoutInflater
                .from(getContext())
                .inflate(R.layout.player_video_progress, this, false);

        seekBarView = (SeekBar) LayoutInflater
                .from(getContext())
                .inflate(R.layout.player_video_seekbar, this, false);

        publishControllerView(progressView);
        publishControllerView(seekBarView);
        updateTimeline();

        seekBarView.setOnSeekBarChangeListener(seekbarChangeListener);
    }

    @Override
    public void onResume() {
        super.onResume();
        updateTimeline();
    }

    @Override
    public void playMedia() {
        super.playMedia();
        updateTimeline();
    }

    @Override
    protected boolean onSingleTap(MotionEvent event) {
        if (userSeekable() && !seekCurrentlyVisible()) {
            logger.info("Show seekbar after tap.");

            lastUserInteraction = System.currentTimeMillis();
            showSeekbar(true);
            return true;
        }

        return super.onSingleTap(event);
    }

    protected boolean userSeekable() {
        return false;
    }

    private boolean seekCurrentlyVisible() {
        return seekBarView.getVisibility() == VISIBLE;
    }

    private void showSeekbar(boolean show) {
        int deltaY = AndroidUtility.dp(getContext(), 12);

        View viewToShow = show ? seekBarView : progressView;
        View viewToHide = show ? progressView : seekBarView;

        if (viewToHide.getVisibility() == VISIBLE) {
            viewToHide.setTranslationY(0);
            viewToHide.animate()
                    .alpha(0)
                    .translationY(deltaY)
                    .setListener(endAction(() -> viewToHide.setVisibility(GONE)))
                    .setInterpolator(new AccelerateInterpolator())
                    .start();
        }


        if (viewToShow.getVisibility() != VISIBLE) {
            viewToShow.setAlpha(0);
            viewToShow.setTranslationY(deltaY);
            viewToShow.setVisibility(VISIBLE);
            viewToShow.animate()
                    .alpha(1)
                    .translationY(0)
                    .setListener(null)
                    .setInterpolator(new DecelerateInterpolator())
                    .start();
        }
    }

    private void updateTimeline() {
        if (!isPlaying())
            return;

        if (!progressTouched) {
            ProgressInfo info = getVideoProgress().orNull();
            if (info != null && shouldShowView(info)) {
                if (firstTimeProgressValue) {
                    firstTimeProgressValue = false;
                    progressView.setVisibility(VISIBLE);
                    progressView.setAlpha(1);
                    progressView.setTranslationY(0);
                }

                for (ProgressBar view : new ProgressBar[]{progressView, seekBarView}) {
                    view.setMax(1000);
                    view.setProgress((int) (1000 * info.progress));
                    view.setSecondaryProgress((int) (1000 * info.buffered));
                }

                if (userSeekable() && seekHideTimeoutReached()) {
                    logger.info("Hiding seekbar after idle timeout");
                    lastUserInteraction = -1;
                    showSeekbar(false);
                }
            } else {
                lastUserInteraction = -1;
                firstTimeProgressValue = true;
                seekBarView.setVisibility(GONE);
                progressView.setVisibility(GONE);
            }
        }

        if (progressEnabled) {
            removeCallbacks(this::updateTimeline);
            postDelayed(this::updateTimeline, 200);
        }
    }

    private boolean seekHideTimeoutReached() {
        return seekCurrentlyVisible()
                && lastUserInteraction > 0
                && System.currentTimeMillis() - lastUserInteraction > 3000;
    }

    private boolean shouldShowView(@NonNull ProgressInfo info) {
        return progressEnabled
                && (info.progress >= 0 && info.progress <= 1
                || info.buffered >= 0 && info.buffered <= 1);
    }

    protected abstract Optional<ProgressInfo> getVideoProgress();

    /**
     * Implement to seek after user input.
     */
    protected void userSeekTo(float fraction) {
    }

    public void setProgressEnabled(boolean visible) {
        progressEnabled = visible;

        if (!visible) {
            lastUserInteraction = -1;
            firstTimeProgressValue = true;
            seekBarView.setVisibility(GONE);
            progressView.setVisibility(GONE);
        }
    }

    protected static class ProgressInfo {
        final float progress;
        final float buffered;

        ProgressInfo(float progress, float buffered) {
            this.buffered = buffered;
            this.progress = progress;
        }
    }

    @SuppressWarnings("FieldCanBeLocal")
    private final SeekBar.OnSeekBarChangeListener seekbarChangeListener = new SeekBar.OnSeekBarChangeListener() {
        @Override
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            if (fromUser) {
                lastUserInteraction = System.currentTimeMillis();
            }
        }

        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {
            progressTouched = true;
            lastUserInteraction = System.currentTimeMillis();
        }

        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {
            int currentValue = seekBar.getProgress();
            userSeekTo(currentValue / (float) seekBar.getMax());

            progressTouched = false;
            lastUserInteraction = System.currentTimeMillis();
        }
    };
}

File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/AbstractProgressMediaView.kt
package com.pr0gramm.app.ui.views.viewer

import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import android.widget.ProgressBar
import android.widget.SeekBar
import androidx.annotation.LayoutRes
import com.pr0gramm.app.Logger
import com.pr0gramm.app.R
import com.pr0gramm.app.util.dip2px
import com.pr0gramm.app.util.visible

/**
 */
@Suppress("LeakingThis")
abstract class AbstractProgressMediaView(config: MediaView.Config, @LayoutRes layoutId: Int?) : MediaView(config, layoutId) {
    internal var progressTouched = false
    internal var lastUserInteraction: Long = -1

    private var progressEnabled = true
    private var firstTimeProgressValue = true

    private val seekBarView: SeekBar = LayoutInflater.from(context)
            .inflate(R.layout.player_video_seekbar, this, false) as SeekBar

    private val progressView: ProgressBar = LayoutInflater.from(context)
            .inflate(R.layout.player_video_progress, this, false) as ProgressBar

    // keep a reference to the update callback so we can re-use
    // the instance to schedule and de-schedule it later.
    private val updateCallback = Runnable { updateTimeline() }

    private var maxBufferedValue = 0.0f

    protected abstract fun currentVideoProgress(): ProgressInfo?

    init {
        publishControllerView(progressView)
        publishControllerView(seekBarView)

        updateTimeline()

        seekBarView.setOnSeekBarChangeListener(SeekbarChangeListener())
    }

    override fun onResume() {
        super.onResume()
        updateTimeline()
    }

    override fun playMedia() {
        super.playMedia()
        updateTimeline()
    }


    override fun onSingleTap(event: MotionEvent): Boolean {
        if (userSeekable()) {
            if (seekCurrentlyVisible()) {
                lastUserInteraction = -1
                showSeekbar(false)
            } else {
                lastUserInteraction = System.currentTimeMillis()
                showSeekbar(true)
            }

            return true
        }

        return super.onSingleTap(event)
    }

    protected open fun userSeekable(): Boolean {
        return false
    }

    private fun seekCurrentlyVisible(): Boolean {
        return seekBarView.visibility == View.VISIBLE
    }

    private fun showSeekbar(show: Boolean) {
        val deltaY = context.dip2px(12)

        val viewToShow = if (show) seekBarView else progressView
        val viewToHide = if (show) progressView else seekBarView

        if (viewToHide.visibility == View.VISIBLE) {
            viewToHide.translationY = 0f
            viewToHide.animate()
                    .alpha(0f)
                    .translationY(deltaY.toFloat())
                    .withEndAction { viewToHide.visible = false }
                    .setInterpolator(AccelerateInterpolator())
                    .start()

        }

        if (viewToShow.visibility != View.VISIBLE) {
            if (progressEnabled || progressView !== viewToShow) {
                viewToShow.alpha = 0f
                viewToShow.translationY = deltaY.toFloat()
                viewToShow.visibility = View.VISIBLE
                viewToShow.animate()
                        .alpha(1f)
                        .translationY(0f)
                        .setListener(null)
                        .setInterpolator(DecelerateInterpolator())
                        .start()
            }

            onSeekbarVisibilityChanged(show)
        }

        if (show) {
            updateTimeline()
        }
    }

    protected open fun onSeekbarVisibilityChanged(show: Boolean) {
        // can be implemented in a subclass.
    }

    private fun updateInfo(view: ProgressBar, info: ProgressInfo) {
        // do not step back in buffering
        val buffered = info.buffered.coerceAtLeast(maxBufferedValue)
        this.maxBufferedValue = buffered

        view.max = 1000
        view.progress = (1000 * info.progress).toInt()
        view.secondaryProgress = (1000 * buffered).toInt()
    }

    private fun updateTimeline() {
        if (!isPlaying)
            return

        if (!progressTouched) {
            val info = currentVideoProgress()
            if (info != null && shouldShowView(info)) {
                if (firstTimeProgressValue && progressEnabled) {
                    firstTimeProgressValue = false
                    progressView.visible = true
                    progressView.alpha = 1f
                    progressView.translationY = 0f
                }

                updateInfo(progressView, info)
                updateInfo(seekBarView, info)

                if (userSeekable() && seekHideTimeoutReached()) {
                    logger.debug { "Hiding seekbar after idle timeout" }
                    lastUserInteraction = -1
                    showSeekbar(false)
                }
            } else {
                lastUserInteraction = -1
                firstTimeProgressValue = true
                seekBarView.visible = false
                progressView.visible = false
            }
        }

        if (progressEnabled || seekCurrentlyVisible()) {
            removeCallbacks(updateCallback)
            postDelayed(updateCallback, 200)
        }
    }

    private fun seekHideTimeoutReached(): Boolean {
        return seekCurrentlyVisible()
                && lastUserInteraction > 0
                && System.currentTimeMillis() - lastUserInteraction > 3000
    }

    private fun shouldShowView(info: ProgressInfo): Boolean {
        return (progressEnabled || seekCurrentlyVisible()) && (info.progress in 0.0..1.0 || info.buffered in 0.0..1.0)
    }

    /**
     * Implement to seek after user input.
     */
    protected open fun userSeekTo(fraction: Float) {
    }

    /**
     * Disable the little progressbar. We still allow seeking, if the user
     * touches the screen.
     */
    fun hideVideoProgress() {
        progressEnabled = false

        lastUserInteraction = -1
        firstTimeProgressValue = true
        seekBarView.visible = false
        progressView.visible = false
    }

    class ProgressInfo(val progress: Float, val buffered: Float)

    private inner class SeekbarChangeListener : SeekBar.OnSeekBarChangeListener {
        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
            if (fromUser) {
                lastUserInteraction = System.currentTimeMillis()
            }
        }

        override fun onStartTrackingTouch(seekBar: SeekBar) {
            progressTouched = true
            lastUserInteraction = System.currentTimeMillis()
        }

        override fun onStopTrackingTouch(seekBar: SeekBar) {
            val currentValue = seekBar.progress
            userSeekTo(currentValue / seekBar.max.toFloat())

            progressTouched = false
            lastUserInteraction = System.currentTimeMillis()
        }
    }

    companion object {
        private val logger = Logger("AbstractProgressMediaView")
    }
}
